workers/handshake-manager/src/manager/internal_engine.rs

```diff
@@ -250,4 +256,33 @@ impl HandshakeExecutor {
 
         Ok((order, wallet))
     }
+
+    /// Check whether a wallet is still valid. This amounts to checking:
+    ///     1. Whether the wallet's known nullifier is still valid. This may be
+    ///        false if the wallet has been updated since a match was attempted
+    ///     2. Whether the wallet's queue is still empty and unpaused.
+    ///        Concurrent matches from elsewhere in the relayer may cause this
+    ///        second condition to be false
+    ///
+    /// This check may be executed after a match settlement fails
+    async fn wallet_still_valid(&self, wallet: &Wallet) -> Result<bool, HandshakeManagerError> {
+        // Check the nullifier
+        let new_wallet = self.state.get_wallet(&wallet.wallet_id).await?;
+        let nullifier = new_wallet.map(|w| w.get_wallet_nullifier()).unwrap_or_default();
+        if wallet.get_wallet_nullifier() != nullifier {
+            return Ok(false);
+        }
+
+        // Check the queue
+        let queue_len = self.state.get_task_queue_len(&wallet.wallet_id).await?;
+        if queue_len > 0 {
+            return Ok(false);
+        }
+
+        if !self.state.is_queue_paused(&wallet.wallet_id).await? {
+            return Ok(false);
+        }
```

Comment:
should this be:
```suggestion
        if self.state.is_queue_paused(&wallet.wallet_id).await? {
            return Ok(false);
        }
```

i.e. if the wallet's queue is now paused, it is no longer valid for matching